content/posts/fission_builder_podspec.md

```diff
@@ -0,0 +1,181 @@
+
+# kubectl label nodes minikube machinecap=high
+node/minikube labeled
+
+# kubectl get pods -n fission-builder 
+NAME                                  READY   STATUS    RESTARTS   AGE
+py-builder-1971419-6dd5658569-c88tc   2/3     Running   0          10m
+```
+
+So, as you can see in the example given we can now provide `podspec` for builder while creating an environment, which enables us a lot of new functionality. 
+
+## Using tolerations in builder podspec 
+Let's take another example, we will now taint the node minikube and then try to provide toleration in the pods to get it successfully scheduled on the minikube node. Set the tain on the node using below command
+```
+# kubectl taint node minikube env=prod:NoSchedule
+node/minikube tainted
+```
+and if we apply the same environment spec once again and check the builder pod that is created, you will be able to see that in `Pending` state
+```
+# kubectl get pods -n fission-builder
+NAME                                  READY   STATUS    RESTARTS   AGE
+py-builder-1973401-6f847c9448-nkg5b   0/3     Pending   0          10s
+```
+and describing it will tell us that the node has taint that this builder pod doesn't tolerate
+```
+# kubectl describe po -n fission-builder py-builder-1973401-6f847c9448-nkg5b
+Events:
+  Type     Reason            Age                 From               Message
+  ----     ------            ----                ----               -------
+  Warning  FailedScheduling  39s (x3 over 115s)  default-scheduler  0/1 nodes are available: 1 node(s) had taints that the pod didn't tolerate.
+
+```
+
+Now let's go ahead and try to create the builder pod with the tolerations set, change the environment spec like the file given below 
+{{< highlight "yaml" >}}
+apiVersion: fission.io/v1
+kind: Environment
+metadata:
+  creationTimestamp: null
+  name: py-builder
+  namespace: default
+spec:
+  builder:
+    command: build
+    image: fission/python-builder
+    container: 
+      command: ["sleep", "1"]
+    podspec:
+      containers:
+      - name: containertwo 
+        image: nginx 
+      tolerations:
+      - key: "env"
+        value: "prod"
+        operator: "Equal" 
+        effect: "NoSchedule"
+      nodeSelector: 
+        machinecap: high 
+  keeparchive: false
+  poolsize: 3
+  resources: {}
+  runtime:
+    image: fission/python-env
+    terminationGracePeriod: 360
+  version: 2
+{{< /highlight >}}
+
+You can see that we are setting `tolerations` for builder pod in the environment spec. Let's apply this environment spec and check if the pod is scheduled on `minikube` node successfully.
+```
+# kubectl get pod -n fission-builder   
+NAME                                  READY   STATUS    RESTARTS   AGE
+py-builder-1974124-865f5fd498-2m9sn   2/3     Running   0          16s
+```
+Describing the pod will let you know that the pod now has been created with the correct tolerations that we have specified for builder in the environment spec.
+
```

Comment:
How about we add a link to the example podspec? That would be very helpful for users.

Reply:
will we have to wait for example `podspec` to be merged?

Reply:
ok, let's merge it first.